### ISSUE TITLE
Does atom value that is not defined in closest Provider fetched from ancestor Providers?

### ISSUE THREAD

**Issue by kozubsky** [pmndrs/jotai#528]
Hello Daishi! First of all, let me thank you for awesome libraries! =)

Well, let's look at this code:
```js
const countAtom = atom(0)
const colorAtom = atom('red')

const Counter = () => {
    const [count, setCount] = useAtom(countAtom)
    const [color, setColor] = useAtom(colorAtom)
    return <div><span style={{color: color}}>{count}</span> <button onClick={() => setCount((c) => c + 1)}>+1</button></div>
}

const Experiment = () => <Fragment>
    <Provider initialValues={[[countAtom, 10], [colorAtom, 'silver']]}>
        <Counter />
        <Provider>
            <Counter />
        </Provider>
        <Provider initialValues={[[countAtom, 20], [colorAtom, 'blue']]}>
            <Counter />
        </Provider>
    </Provider>
</Fragment>;
```

It will render:

```
10 [+1] // silver (initial value defined in closest provider)
0 [+1] // red (default value of atom) — can inherit silver from parent Provider?
20 [+1] // blue (initial value defined in closest provider)
```

The question is: is there way for counter 2 to fetch ("inherit") a `colorAtom` value from a root provider in case when `colorAtom` is not defined in initialValues of closest provider A?

I am wondering how I can use jotai atoms for using in subtree components with own Provider (for speciefic subtree atoms) alongside with atoms defined in parent scope.

In React Context this is possible by wrapping subtree into < SubtreeContextProvider >, that is being wrapped into some < AppContextProvider >. But React Context is verbose and not so cool in performance.

Any ideas? )

**Comment by dai-shi** [comment]
No, this is not possible by design. It will look for a store in the closest Provider, and that's it.

How is this possible with React Context? Anyway, it will never happen as it would be bad in performance as you imply.

I'm not sure about your use case. We have atom "scope" but I guess it doesn't help you.

**Comment by kozubsky** [comment]
> No, this is not possible by design. It will look for a store in the closest Provider, and that's it.
> 
> How is this possible with React Context? Anyway, it will never happen as it would be bad in performance as you imply.
> 
> I'm not sure about your use case. We have atom "scope" but I guess it doesn't help you.

Hello Daishi! Thank you for answer!

I understood that atom values "inheritance" is not possible by design in jotai. Also, I realized that "inheritance" in React Context I meant is not inheritance, but other thing.

In React we have createContext function, and we able to create different context "types". For example, we can create `ColorContextProvider`, containing values {color: (default) 'red'} and `CountContextProvider`, containing values {count: (default) 0}. Yes, they are also isolated, but per-type. And, if we mix them, we can get desired behaviour. Something like this:

```js
import React, {createContext, FunctionComponent, useContext, useState} from "react";
import { Fragment } from "react";
import {render} from "react-dom";

interface ICountContextData {count: number};
const CountContext = createContext<[ICountContextData, any]>([{count: 0}, () => {}]);
const CountContextProvider: FunctionComponent<Partial<ICountContextData>> = (props) => {
    const [state, setState] = useState<ICountContextData>({
        count: props.count ?? 0,
    });

    return <CountContext.Provider value={[state, setState]}>
        {props.children}
    </CountContext.Provider>;
};

interface IColorContextData {color: string};
const ColorContext = createContext<IColorContextData>({color: 'red'});
const ColorContextProvider: FunctionComponent<Partial<IColorContextData>> = (props) => {
    const [state, setState] = useState<IColorContextData>({
        color: props.color ?? 'red',
    });

    return <ColorContext.Provider value={state}>
        {props.children}
    </ColorContext.Provider>;
};

const ReactCounter = () => {
    const {color} = useContext(ColorContext);
    const [{count}, setCount] = useContext(CountContext);

    return <div>
        <span style={{color: color}}>{count}</span>
        <button onClick={() => setCount({count: count + 1})}>+1</button>
    </div>
}

export const ReactExperiment = () => <Fragment>
    <ColorContextProvider color={'silver'}>
        <CountContextProvider>
            <ReactCounter />
        </CountContextProvider>
        <ColorContextProvider color={'blue'}>
            <CountContextProvider count={20}>
                <ReactCounter />
            </CountContextProvider>
        </ColorContextProvider>
    </ColorContextProvider>
</Fragment>;


render(<ReactExperiment />, document.getElementById('root'));
```

That will result in:

```
0 [+1] // silver (value defined in closest AContextProvider provider — first)
20 [+1] // blue (value defined in closest AContextProvider provider - second)
```

But... Such solution based on React Context seems is bad for performance. I looked for alternative solution. But in Jotai there is only one Provider "type", as I know. Or scopes can help with that (is not it something like "type")?

P.S. It is very simplified example, real case is multilevel tree of UI components, and on some ("middle") levels I need to share some state down to "leafs". But I need to provide not only that "middle" level state: also I need to provide to descendants data that was shared on "top" level, with data mutators. Changes of that "top" level data should take effect on all components starting from "top" level, not subtree only. For example, 4 editable image gallereys on same page with different initial images on the right and multilevel tree with editable nodes on the left.

In other words, tree of visual components and provider components can be:

```
root
[ROOT data provider]: top1, ..., topN
  component A
    [A data provider]: a1, ..., aN
      component B
        component C
          [C data provider]: a1, ..., aN
            component D (can use all layers data: top1, a4, a6, c10. Not only closest C provider)
```

**Comment by kozubsky** [comment]
Wow, just tested jotai Providers scoping, and it works, the behaviour seems to be the same as in React Context example above.

```js
import {atom, Provider, useAtom} from "jotai";
import React, {Fragment} from "react";

const countScope = Symbol();
const countAtom = atom(0)
countAtom.scope = countScope;

const colorScope = Symbol();
const colorAtom = atom('red')
colorAtom.scope = colorScope;

export const Counter = () => {
    const [count, setCount] = useAtom(countAtom)
    const [color, setColor] = useAtom(colorAtom)
    return <div>
        <span style={{color: color}}>{count}</span>
        <button onClick={() => setCount((c) => c + 1)}>+1</button>
    </div>
}

export const JotaiExperiment2 = () => <Fragment>
    <Provider scope={countScope}>
        <Provider scope={colorScope} initialValues={[[colorAtom, 'silver']]}>
            <Counter />
            <Provider scope={colorScope} initialValues={[[colorAtom, 'blue']]}>
                <Provider scope={countScope} initialValues={[[countAtom, 20]]}>
                    <Counter />
                </Provider>
            </Provider>
        </Provider>
    </Provider>
</Fragment>;

```
Can you overview this solution from performance point of view? Should it be better than with React Context solution? What if we increase Providers count to 50, for example?

**Comment by dai-shi** [comment]
Good, so it seems like scoping works for you. Internally, it's just multiple React contexts, and the performance should be the same (meaning good).

**Comment by dai-shi** [comment]
Closing as answered. Let us know if you have more questions.
